### ISSUE TITLE
Diffrent output for the same query run on server and client

### ISSUE THREAD

**Issue by elisei-timofte** [meteor/meteor#9111]
- [ ] The version of Meteor showing the problem: METEOR@1.5.2
- [ ] The last version of Meteor where the problem did _not_ occur: METEOR@1.4.3.2
- [ ] The operating system you're running Meteor on: OSX Sierra 10.12.6
- [ ] The expected behavior: The query should return the same thing on the server and client
- [ ] The actual behavior: The server returns nothing (as expected) while the client return an object
- [ ] A **simple** reproduction!: https://github.com/elisei-timofte/meteor-mongo-bug

**Comment by hwillson** [comment]
Thanks for the excellent reproduction @elisei-timofte! Confirming as a bug.

**Comment by jamiter** [comment]
@hwillson, you marked it as having a work-around. What is the work-around?

We rely heavily on $elemMatch queries, so this is a high prio issue for us. Or we need to downgrade to 1.5, if that's still working.

**Comment by elisei-timofte** [comment]
I'm also forced to delay deploying into production the new version until this gets fixed.

**Comment by hwillson** [comment]
@jamiter I marked this initially as having a work around as in some cases `$not` + `$elemMatch` queries can be re-written to leverage `$ne` instead of `$not` (which is often done to work around issues with Mongo indexes and `$not`). That being said, I understand re-writing queries is not a great workaround. I'll bump this up to `Severity:production`.

With regards to this issue, I have tracked down the problem that's breaking the current reproduction. I've been working on a PR but have unfortunately uncovered more related issues that are taking longer to fix. Not only do queries like the following fail:

```js
// Collection data stored in Mongo
{
  dogs: [
    { name: 'Rex', age: 3 }
  ]
}

// Case 1: Minimongo selector that shouldn't return a match but does
{
  $or: [
    { dogs: { $not: { $elemMatch: { name: 'Rex' } } } },
    { dogs: { $elemMatch: { name: 'Rex', age: 5 } } }
  ]
}
```

but so do queries like:

```js
// Case 2: Minimongo selector that should match but doesn't
{
  $or: [
    { dogs: { $elemMatch: { name: 'Rex' } } },
    { dogs: { $elemMatch: { name: 'Rex', age: 5 } } }
  ]
}
```

These problems are being caused by separate reasons. `Case 1` (which lines up with the repro) is being caused because its first `$elemMatch` call when run by itself is returning `false` when it should be returning `true`, at which point that `false` is inverted to `true` by the `$not` and a match is returned. `Case 2` however appears to be caused by the code behind the `$or` operator re-using code that is intended for the `$and` operator; when debugging I can see that each individual `dogs` selector in `Case 2` when run by itself is working (first one returns `true`, second one returns `false`), but then the code behind the `$or` operator is wrongly expecting all sub-selectors in this case to return `true` (since it's re-using parts of the `$and` code).

Long story short, I'm still working on this and hope to have more details shortly.

**Comment by quintstoffers** [comment]
This also seems to affect observing changes server-side, here's the code we used to reproduce:
https://gist.github.com/quintstoffers/ec49c37af7f48f5695ff060934593eba

After digging through Minimongo's code this seems to be caused by a missing variable declaration for `subMatcher`:
https://github.com/meteor/meteor/blob/devel/packages/minimongo/common.js#L182

Essentially `subMatcher` is global, so the last call to this function sets the subMatcher that is used for all element selectors.

**Comment by hwillson** [comment]
Nice find @quintstoffers! Looks like it was dropped in

https://github.com/meteor/meteor/commit/fe576f60ce019212699af53660d2afbcb3a721e6#diff-16b62555146b850d7aa9e8aa8e5b8102. 

Submitting a PR with additional tests for this momentarily. Thanks!

**Comment by elisei-timofte** [comment]
Should this be working now? It this fix is included in the new version 1.5.2.1?

**Comment by hwillson** [comment]
@elisei-timofte Yes, the fix is in 1.5.2.1.

**Comment by elisei-timofte** [comment]
There is still an [issue](https://github.com/meteor/meteor/issues/9136) with 1.5.2.1 but I'll wait for the fix. 

Thanks again for your help!
